### DISCUSSION TITLE
Jvw.DevToys.SemverCalculator extension

### DISCUSSION THREAD

**Discussion by jerone** [DevToys-app/DevToys#1333] [Show and tell]
First of all; awesome tool. Very useful and solid tooling. Also love that it's build in .NET.

# Extension
I wanted to share my own extension to DevToys: **Jvw.DevToys.SemverCalculator**

It's a Semantic Versioning (SemVer) calculator for NPM packages with the official v2.0.0 spec from [semver.org](http://semver.org/).

Download here: https://www.nuget.org/packages/Jvw.DevToys.SemverCalculator
Source code: https://github.com/jerone/Jvw.DevToys.SemverCalculator

![image](https://github.com/user-attachments/assets/d5e0607b-5347-44f2-90ca-62c731b77757)

# Technical
The extension makes a HTTP request to registry.npmjs.org to fetch the package information. This contains a list of all versions ever released by that package. The extension extracts this list of versions and shows it on the screen. 
It then takes the semver range input value and shows an indicator on all versions that match the range.

# Learnings
The documentation was clear and complete. Super helpful. Especially the UI examples WITH screenshots.

I did had troubles using the environment variables `DevToysGuiDebugEntryPoint` and `DevToysCliDebugEntryPoint`, because they point to the preview version of the application, which is installed in a folder with a space in it "DevToys Preview". For some reason the `launchSettings.json` in VS cannot handle this. I tried with extra quotes, but no success. For now I [hard-coded](https://github.com/jerone/Jvw.DevToys.SemverCalculator/blob/master/Jvw.DevToys.SemverCalculator/Properties/launchsettings.json#L6) the paths.

During packaging, I initially skipped over the `IncludeAllFilesInTargetDir` step because I thought I did not need this. But my project was referencing another NuGet package (Semver), so I did need this step. The packaging step was complaining about files that were already there, plus it was copying over other `.dll` that should not be published (CSharpier). So, I changed to code to [explicitly include only the `.dll` file that I needed](https://github.com/jerone/Jvw.DevToys.SemverCalculator/blob/master/Jvw.DevToys.SemverCalculator/Jvw.DevToys.SemverCalculator.csproj#L30-L37).

It was my first official full experience with making an NuGet package, and with the documentation this was very easy.

All UI components were easy to work with. For my use-case I missed some styling feature, like text and border color options, but I solved it with emoji's.

I noticed that it was possible to have your own extension installed and sequentially run the extension via debug. It's a little hard to distinct between the 2 versions, but this could be solved with a `#if DEBUG` around the extension name.

**Comment by jerone** [comment]
I noticed [the documentation](https://devtoys.app/doc/articles/extension-development/guidelines/use-settings.html?tabs=windows) mentions to create an unique name for settings by using the tool name.

```csharp
name: $"{nameof(MyGuiTool)}.{nameof(mySetting)}", // Unique name for the setting. Use the tool name to avoid conflicts.
```

However when I go to the `settings.ini` file, I see all settings are already prefixed with the tool name.

For example, I created a setting called "HttpAgreed": 
```csharp
    public static SettingDefinition<bool> HttpAgreed =
        new(name: "Jvw.DevToys.SemverCalculator." + nameof(HttpAgreed), defaultValue: false);
```
And this appeared in the `settings.ini` file:
```ini
Jvw.DevToys.SemverCalculator.Jvw.DevToys.SemverCalculator.HttpAgreed=True
```

Is the suggestion in the docs still relevant?

**Reply by veler** [reply]
Hello, thank you for all these feedback! This is super helpful. Also good catch on this mistake :) I will correct it.

**Comment by jerone** [comment]
@veler Is it correct that if I have my own extension installed in DevToys, and I start a debug session running Visual Studio with my extension, that both extensions (installed-before and debug) referrer to the debug-session?

I expected them to not be the same.

I tried changing the `Name`-attribute to something different.
I also [changed the `ShortDisplayTitleResourceName`- & `LongDisplayTitleResourceName`-attributes](https://github.com/jerone/Jvw.DevToys.SemverCalculator/blob/master/Jvw.DevToys.SemverCalculator/Gui.cs#L23-L29) to distinct between the two versions.

**Reply by veler** [reply]
> Is it correct that if I have my own extension installed in DevToys, and I start a debug session running Visual Studio with my extension, that both extensions (installed-before and debug) referrer to the debug-session?

That looks like a bug (I just reproduced it). Tools from both extensions appear in the navigation bar, and clicking both show the debugged one. Ideally, it should only show the one being debugged, I think.

**Reply by jerone** [reply]
Yes, agree.

**Reply by jerone** [reply]
@veler Where you able to find a solution for this bug?

I had to uninstall the manual installation of my extension, because this was conflicting with MEF `ImportingConstructor`-attribute trying to import duplicated services, as it would find both the manual extension service and the debug extension service.

**Comment by jerone** [comment]
@veler As you can see in my screenshot above, the list of results can be very long. I couldn't find any UI element that had a scrollable option (besides the screen itself). Would it be possible to add a `max-height` option to a `UICard` (of other element)?

**Reply by veler** [reply]
Did you try `List`? It has a scroll bar when it gets too large: https://devtoys.app/doc/articles/extension-development/guidelines/UX/collections/list.html

That's what Extension Manager uses: https://github.com/DevToys-app/DevToys/blob/ee85fda03c33192988d2be49c16aa5760508f000/src/app/dev/DevToys.Blazor/BuiltInTools/ExtensionsManager/ExtensionsManagerGuiTool.cs#L131-L132

Same thing with PNG Compressor and Image Converter.

**Reply by jerone** [reply]
I was talking about this element 👇 This would be not a great UX when all versions are rendered in a `List`:
![image](https://github.com/user-attachments/assets/a5ebddd9-9525-4063-a1a6-3a843ac15952)

Right now, I'm adding a cheatsheet `DataGrid` on the right. But the `DataGrid` is taking a height of the full window, instead of the content. Creating a scrollbar (see purple) with only empty space (see green).
![image](https://github.com/user-attachments/assets/74ee5281-f566-4d4f-8c6f-cbfb26e6f144)

When I disable the `isScrollable` on the `UIToolView`, the cheatsheet `DataGrid` is rendered correctly, but the results `UICard` get clipped (unscrollable) when the data is more than the available space.

**Reply by veler** [reply]
I see. Right now, WrapPanel is not meant to auto scroll like you seem to wish. Perhaps I can add a "ScrollContainer" control. Just an idea...

**Reply by jerone** [reply]
I think that would be a great addition.

**Comment by jerone** [comment]
Just wanted to share that I was able to [unit-test](https://github.com/jerone/Jvw.DevToys.SemverCalculator/tree/master/Jvw.DevToys.SemverCalculator.Tests) this whole extension. Including the [GUI](https://github.com/jerone/Jvw.DevToys.SemverCalculator/blob/master/Jvw.DevToys.SemverCalculator.Tests/Tests/Programs/GuiTests.cs).

I build an custom [converter](https://github.com/jerone/Jvw.DevToys.SemverCalculator/blob/master/Jvw.DevToys.SemverCalculator.Tests/Converters/DevToysElementConverter.cs) for [Verify](https://github.com/VerifyTests/Verify) to allow for snapshot testing of the UI elements.
Example [test](https://github.com/jerone/Jvw.DevToys.SemverCalculator/blob/master/Jvw.DevToys.SemverCalculator.Tests/Tests/Components/CheatSheetComponentTests.cs) & [snapshot](https://github.com/jerone/Jvw.DevToys.SemverCalculator/blob/master/Jvw.DevToys.SemverCalculator.Tests/Tests/Components/CheatSheetComponentTests.CheatSheetComponent_HasRows.verified.txt).

**Reply by veler** [reply]
Oh, this is awesome! :D

**Comment by jerone** [comment]
I'm experiencing a weird behavior with one particular extension NuGet package; `NuGet.Versioning`.
After adding it to my extension, the extension will **not** load at all anymore.

To confirm that the issue is not related to the content, I tried a similar extension (`Chocolatey.NuGet.Versioning`) without any issues.
I also included the source code from `NuGet.Versioning` directly inside my extension without any issue.

The output console mentions nothing interesting.
But I did saw there was already a dll being loaded called `NuGet.Versioning.dll`.
After some investigation, I noticed that DevToys [imports `NuGet.Packaging`](https://github.com/DevToys-app/DevToys/blob/a8f22e522727fe8ae91e06ac81ad42829390fcad/src/Directory.Packages.props#L34), which has a [dependency on `NuGet.Versioning`](https://www.nuget.org/packages/NuGet.Packaging#dependencies-body-tab).

**@veler Do you know how I can still use `NuGet.Versioning` package in my extension?**

_This could be an issue with any other packages that DevToys includes, but I have not confirmed that._

**Reply by veler** [reply]
Hi,
This is an interesting issue and was worried it would happen at some point. As you found out, DevToys has `NuGet.Versioning` as a dependency, hence, it's already loaded.

Unfortunately, [.NET does not support `AppDomain`](https://stackoverflow.com/a/45600956) like .NET Framework 4.7.2 does. AppDomains would be a good way to solve this issue : we could make 1 AppDomain per extension. Each AppDomain can load its own set of assemblies independently.

Potential alternative: Out-of-proc model. Each extension run in its own program independently. So on Windows for example, imagine that each extension is its own .exe running in parallel of DevToys.exe.

Pros:
1. Extensions are isolated. No dependencies issues anymore.
2. When an extension crash, it doesn't crash the rest of the app.

Cons:
1. This is a big breaking change
2. It will be a nightmare for handling UI

Anyway... With this being said, let's explore some alternative to extension isolation:
1. Use binding redirect in DevToys.
2. Your extension could load NuGet.Versioning dynamically maybe (not ideal for long term solution, for sure).
3. DevToys to use [`AssemblyLoadContext`](https://learn.microsoft.com/en-us/dotnet/core/dependency-loading/understanding-assemblyloadcontext). Note that I just learnt about it while exploring solutions for your problem. I'm also not sure how it would work with MEF. Worth a try either way I think.

**Reply by jerone** [reply]
> This is an interesting issue and was worried it would happen at some point. As you found out, DevToys has `NuGet.Versioning` as a dependency, hence, it's already loaded.

With almost 400 dll files in the install folder, the change is relatively high it might happen again.

I tried excluding `NuGet.Versioning` to the output folder, but no luck.

> Anyway... With this being said, let's explore some alternative to extension isolation

Sadly I'm not familiar with MEF to give any meaningful advice.
But I'm here if you want to try something.

For now I copied the source code of `NuGet.Versioning` into my extension, because the cloned one by Chocolatey was not signed anyways.

**Reply by jerone** [reply]
Haven't tried this, but would it make any difference if I reference the same version of `NuGet.Versioning` package?

I was constantly trying with the latest version, but I just noticed that DevToys is using an older version 6.9.1.

**Reply by veler** [reply]
Using the same version should work. But it isn't a future-proof solution as I could update this package anytime and we would end up in the same situation again.

**Reply by veler** [reply]
Hi @jerone , I created an issue to track this problem #1430
And well... I also found a fix using `AssemblyLoadContext` :D Giving credits to GitHub Copilot for suggesting this API and providing the code change.
#1431 
In the PR description, there's a ZIP that contains a test extension you can use to verify whether it address the issue. Feel free to try with your own extension too :)

**Reply by jerone** [reply]
I tried it and it works great. 👍



![](https://programmerhumor.io/wp-content/uploads/2023/04/programmerhumor-io-linux-memes-backend-memes-b2c93c486cdc54d.png)

**Comment by veler** [comment]
Thanks :) If I may, I will close this discussion for now.
